### PR TITLE
Notification string to a queue of strings

### LINKED ISSUE #2469: Rework command enqueing to minimize swallowing
## Is your feature request related to a problem? Please describe.

It is currently not possible to call `m_sig.emit(button_press{*});` more than once before `m_inputdata` is cleared in the next iteration of the eventqueue worker loop. This is a problem, for example, if more than one action need to be triggered in the same method call.

## Why does polybar need this feature?

This feature would make the core system underlying polybar considerably more flexible, and would allow for easier and faster implementation of some features.

## Describe the solution you'd like

Instead of `m_inputdata` there could be a queue of input data strings.

## Describe alternatives you've considered

One alternative is to not emit many signals one after the other, and instead store the signals and call them after a short delay, but this increases complexity and makes everything more prone to bugs.

## Additional context

image
Results in
image

### PR DESCRIPTION
**Pull request by madhavpcm**
## What type of PR is this? (check all applicable)

* [ ] Refactor
* [ ] Feature
* [ ] Bug Fix
* [x] Optimization
* [ ] Other: *Replace this with a description of the type of this PR*

## Related Issues & Documents

Fixes #2469
## Documentation (check all applicable)

* [ ] This PR requires changes to the Wiki documentation (describe the changes)
* [ ] This PR requires changes to the documentation inside the git repo (please add them to the PR).
* [x] Does not require documentation changes

### CODE REVIEW

File: include/common.hpp [hunk 1]
```diff
@@ -35,7 +36,7 @@ using std::make_pair;
 using std::array;
 using std::vector;
 using std::to_string;
-
+using std::queue;
```
Comment (patrick96) [inline include/common.hpp:39]: Please remove this. We don't necessarily want to force everyone to include the `queue` header if they don't need it.

File: src/components/controller.cpp [hunk 2]
```diff
@@ -80,7 +80,7 @@ controller::~controller() {
 }
 
 /**
- * Run the main loop
+ * Run the main loopParameters
```
Comment (patrick96) [inline src/components/controller.cpp:83]:
```suggestion
 * Run the main loop
```

File: src/components/controller.cpp [hunk 3]
```diff
@@ -215,13 +211,17 @@ void controller::notifier_handler() {
     return;
   }
 
-  if (!data.inputdata.empty()) {
-    process_inputdata(std::move(data.inputdata));
+  while(!data.inputdata.empty()) {
+    m_log.trace("controller: Dequeueing inputdata : %s ",data.inputdata.front());
+    auto inputdata = data.inputdata.front();
+    data.inputdata.pop();
+    process_inputdata(std::move(inputdata));
```
Comment (patrick96) [inline src/components/controller.cpp:218]:
```suggestion
    auto inputdata = data.inputdata.front();
    data.inputdata.pop();
    m_log.trace("controller: Dequeueing inputdata : %s ", inputdata);
    process_inputdata(std::move(inputdata));
```

File: src/components/controller.cpp [hunk 4]
```diff
@@ -128,13 +128,9 @@ bool controller::run(bool writeback, string snapshot_dst, bool confwatch) {
...
 void controller::trigger_action(string&& input_data) {
   std::unique_lock<std::mutex> guard(m_notification_mutex);
-
-  if (m_notifications.inputdata.empty()) {
-    m_notifications.inputdata = std::move(input_data);
-    trigger_notification();
-  } else {
-    m_log.trace("controller: Swallowing input event (pending data)");
-  }
+  m_log.trace("controller: Queueing input event %s %d", input_data,input_data.size());
```
Comment (patrick96) [inline src/components/controller.cpp:131]:
```suggestion
  m_log.trace("controller: Queueing input event '%s'", input_data);
```

File: src/components/controller.cpp [hunk 5]
```diff
@@ -233,58 +233,62 @@ void controller::screenshot_handler() {
...
+            // Trigger a single screenshot after 3 seconds
+            m_loop->timer_handle([this]() { screenshot_handler(); })->start(3000, 0);
+        }
 
-    m_notifier = m_loop->async_handle([this]() { notifier_handler(); });
+        m_notifier = m_loop->async_handle([this]() { notifier_handler(); });
+       /* m_loop->timer_handle([this]() {
+            trigger_action("notify-send A");
+            trigger_action("notify-send B");
+        })->start(500,0); */
```
Comment (patrick96) [inline src/components/controller.cpp:269]: Please remove

**Review by patrick96 — CHANGES_REQUESTED** [review]
Changes look good :+1: 

I commented mainly on minor formatting issues. Please also [format your code](https://github.com/polybar/polybar/wiki/Style-Guide#code-formatting) with `clang-format`.

An easy way to format all changed files in the PR is by running in the root directory of the repo:

```bash
git show --pretty="" --name-only "$(git merge-base HEAD upstream/master)..HEAD" | xargs clang-format -i --style=file
```

This will also format non-c++ files, just take care to not add those changes to git because `clang-format` can mess those up pretty badly.

And finally, please mention your changes in `CHANGELOG.md` in the `Changed` section: https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md#changelog

File: include/common.hpp (hunk 1 above)
Comment (madhavpcm) [reply include/common.hpp:39]: Is it okay if I moved the include queue to controller.hpp? or just the using std::queue;

File: src/components/controller.cpp (hunk 5 above)
Comment (madhavpcm) [reply src/components/controller.cpp:269]: I'm sorry about the formatting issues, I have a couple of binds that changes the indentation of selected code.

**Review by patrick96 — APPROVED** [review]
Looks good now.

I also did some minor cleanup.

Thanks for your efforts :smiley:

**Pull request merged**